executable-monitor/executable-monitor.py

```diff
@@ -72,76 +81,114 @@
+
+        # Create two file descriptors. The subprocess writes to one, the parent task reads from the other
+        # This is a workaround to avoid the fact that calling readline() on the stdout of the subprocess is
+        # a blocking call. Where if the subproces is running but hasn't printed anything, readline will never time out.
+        # The approach uses the underlying file system to rely on an end of file character to ensure that it does not hang.
+        WriteOutputFile = open("output.log", "w")
+        ReadOutputFile = open("output.log", "r")
+
+        # Launch the executable
+        exe = subprocess.Popen([exe_abs_path], stdout=WriteOutputFile, stderr=WriteOutputFile, universal_newlines=True)
 
-    # Initialize values
-    timeout_occurred = False
-    exe_exit_status = None
-    exe_exitted = False
-    success_line_found = False
-    wait_for_exit = args.success_exit_status is not None
-
-    # Launch the executable
-    exe = subprocess.Popen([exe_abs_path], stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
-
-    cur_time_seconds = time.time()
-    timeout_time_seconds = cur_time_seconds + args.timeout_seconds
-
-    logging.info("START OF DEVICE OUTPUT\n")
-
-    while not (timeout_occurred or exe_exitted or (not wait_for_exit and success_line_found)):
-
-        # Check if executable exitted
-        exe_exit_status = exe.poll()
-        if exe_exit_status is not None:
-            exe_exitted = True
-
-        # Read executable's stdout and write to stdout and logfile
-        exe_stdout_line = exe.stdout.readline()
-        logging.info(exe_stdout_line)
-
-        # Check if the executable printed out it's success line
-        if args.success_line is not None and args.success_line in exe_stdout_line:
-            success_line_found = True
-
-        # Check for timeout
         cur_time_seconds = time.time()
-        if cur_time_seconds >= timeout_time_seconds:
-            timeout_occurred = True
-
-        # Sleep for a short duration between loops to not steal all system resources
-        time.sleep(.1)
-
-    if not exe_exitted:
-        exe.kill()
-
-    # Capture remaining output and check for the successful line
-    for exe_stdout_line in exe.stdout.readlines():
-        logging.info(exe_stdout_line)
-        if args.success_line is not None and args.success_line in exe_stdout_line:
-            success_line_found = True
-    
-    logging.info("END OF DEVICE OUTPUT\n")
-
-    logging.info("EXECUTABLE RUN SUMMARY:\n")
+        timeout_time_seconds = cur_time_seconds + args.timeout_seconds
+
+        logging.info("START OF DEVICE OUTPUT\n")
+
+        # While a timeout hasn't happened, the executable is running, and an exit condition has not been met
+        while ( not exit_condition_met ):
+            # Uncomment this sleep for a short duration between loops to not steal all system resources
+            # time.sleep(.05)
```

Comment:
Does this need to be uncommented? Is the call to readline() blocking?

Reply:
I left it in intentionally for if running it locally if somebody saw it more or less starving the other tasks on the system, such as the subprocess being run.